File: src/lowlevel/server/icefsoperators/icefsFsync.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-17 16:42:07
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsFsync.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
	"syscall"
)

func (s *IcefsServer) doIcefsFsync(dataSync int32, fh uint64) (status int32) {
	var err error
	if dataSync != 0 {
		err = syscall.Fdatasync(int(fh))
	} else {
		err = syscall.Fsync(int(fh))
	}
	status = icefserror.IcefsStdErrno(err)
	return
}

func (s *IcefsGRpcServer) DoIcefsFsync(ctx context.Context, req *pb.IcefsFsyncReq) (*pb.IcefsFsyncRes, error) {
	var res pb.IcefsFsyncRes

	res.Status = s.server.doIcefsFsync(req.DataSync, req.Fh)

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsFsync(ctx context.Context, req *icefsthrift.IcefsFsyncReq) (*icefsthrift.IcefsFsyncRes, error) {
	var res icefsthrift.IcefsFsyncRes

	res.Status = s.server.doIcefsFsync(req.DataSync, uint64(req.Fh))

	return &res, nil
}

File: src/lowlevel/client/icefsoperators/icefsFsync.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:40:45
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsFsync.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsFsync(fuse_req_t fuseReq, fuse_ino_t inode,
                               int dataSync, struct fuse_file_info *fi) {
  ICEFS_PR_FUNCTION;

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsFsyncReq req;
      icefsgrpc::IcefsFsyncRes res;
      grpc::ClientContext ctx;

      req.set_data_sync(dataSync);
      req.set_fh(fi->fh);

      grpc::Status status = gRpcClient->DoIcefsFsync(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        fuse_reply_err(fuseReq, ICEFS_EOK);
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsFsyncReq req;
      icefsthrift::IcefsFsyncRes res;

      req.__set_data_sync(dataSync);
      req.__set_fh(fi->fh);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsFsync(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      fuse_reply_err(fuseReq, res.status);
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}